plugins/org.eclipse.n4js/src/org/eclipse/n4js/postprocessing/PolyProcessor.xtend

```diff
@@ -154,12 +163,27 @@ package class PolyProcessor extends AbstractPolyProcessor {
 		// (until the expectedType judgment is integrated into AST traversal, we have to invoke this judgment here;
 		// in case of not-well-behaving expectedType rules, we use 'null' as expected type, i.e. no expectation)
 		// TODO integrate expectedType judgment into AST traversal and remove #isProblematicCaseOfExpectedType()
-		val expectedTypeRef = if (!rootPoly.isProblematicCaseOfExpectedType) {
+		var expectedTypeRef = if (!rootPoly.isProblematicCaseOfExpectedType) {
 				ts.expectedTypeIn(G, rootPoly.eContainer(), rootPoly).getValue();
 			};
 
+		// In case of destructure pattern, we can calculate the expected type based on the structure of the destructure pattern.
+		var rootDestructNode = if (rootPoly.eContainer instanceof VariableBinding) {
+			DestructNode.unify(rootPoly.eContainer as VariableBinding)
+		} else if (rootPoly.eContainer instanceof AssignmentExpression) {
+			DestructNode.unify(rootPoly.eContainer as AssignmentExpression)
+		} else if (rootPoly.eContainer instanceof ForStatement) {
+			DestructNode.unify(rootPoly.eContainer as ForStatement)
+		}
+		else {
+			null
+		};
```

Comment:
move this also to a helper method `calculateExpectedType(Expression)` (in DestructureHelper)